cpp/src/io/parquet/experimental/deletion_vectors.cu

```diff
@@ -0,0 +1,225 @@
+
+  // Segmented inclusive scan to compute the rest of the row indices
+  thrust::inclusive_scan_by_key(rmm::exec_policy_nosync(stream),
+                                row_group_keys.begin(),
+                                row_group_keys.end(),
+                                row_indices_iter,
+                                row_indices_iter);
+
+  return std::make_unique<cudf::column>(cudf::data_type{cudf::type_id::UINT64},
+                                        num_rows,
+                                        std::move(row_indices),
+                                        rmm::device_buffer{},
+                                        0);
+}
+
+/**
+ * @brief Builds a BOOL8 row mask column from the specified host span of row indices and the
+ * roaring64 deletion vector
+ *
+ * @param row_indices Host span of row indices
+ * @param deletion_vector Pointer to the roaring64 deletion vector
+ * @param num_rows Number of rows in the column
+ * @param stream CUDA stream for kernel launches and data transfers
+ * @param mr Device memory resource to allocate device memory for the row mask column
+ *
+ * @return Unique pointer to the row mask column
+ */
+std::unique_ptr<cudf::column> build_row_mask_column(
+  cudf::column_view const& row_index_column,
+  cudf::host_span<cuda::std::byte const> deletion_vector,
+  size_type num_rows,
+  rmm::cuda_stream_view stream,
+  rmm::device_async_resource_ref mr)
+{
+  auto row_mask = rmm::device_buffer(num_rows, stream, mr);
+  cuco::experimental::roaring_bitmap<cuda::std::uint64_t> roaring_bitmap(
+    deletion_vector.data(), {}, stream);
+
+  // Iterator to negate and store the output value from `contains_async`
+  auto row_mask_iter = thrust::make_transform_output_iterator(
+    static_cast<bool*>(row_mask.data()), [] __device__(auto b) { return not b; });
+  roaring_bitmap.contains_async(
+    row_index_column.begin<size_t>(), row_index_column.end<size_t>(), row_mask_iter, stream);
+
+  return std::make_unique<cudf::column>(
+    cudf::data_type{cudf::type_id::BOOL8}, num_rows, std::move(row_mask), rmm::device_buffer{}, 0);
+}
+
+}  // namespace
+
+table_with_metadata read_parquet_and_apply_deletion_vector(
+  parquet_reader_options const& options,
+  cudf::host_span<cuda::std::byte const> serialized_roaring64,
+  cudf::host_span<size_t const> row_group_offsets,
+  cudf::host_span<size_type const> row_group_num_rows,
+  rmm::cuda_stream_view stream,
+  rmm::device_async_resource_ref mr)
+{
+  CUDF_EXPECTS(
+    row_group_offsets.size() == row_group_num_rows.size(),
+    "Encountered a mismatch in the number of row group offsets and row group row counts");
+  CUDF_EXPECTS(
+    not options.get_filter().has_value(),
+    "Encountered a non-empty AST filter expression. Use a roaring64 bitmap deletion vector to "
+    "filter the table instead");
+
+  // Read the parquet table
+  auto [table, metadata] = cudf::io::read_parquet(options, stream, mr);
```

Comment:
we should use the user-provided mr only when the bitmap is empty

Reply:
Huh, I see what you mean here. Good catch

Reply:
Updated at both places